src/virtualenv/run.py

```diff
@@ -0,0 +1,190 @@
+from __future__ import absolute_import, unicode_literals
+
+import logging
+from argparse import ArgumentTypeError
+
+from entrypoints import get_group_named
+
+from .config.cli.parser import VirtualEnvConfigParser
+from .report import LEVELS, setup_report
+from .session import Session
+from .version import __version__
+
+
+def run_via_cli(args):
+    """Run the virtual environment creation via CLI arguments
+
+    :param args: the command line arguments
+    :return: the creator used
+    """
+    session = session_via_cli(args)
+    session.run()
+    return session
+
+
+def session_via_cli(args):
+    parser = VirtualEnvConfigParser()
+    add_version_flag(parser)
+    options, verbosity = _do_report_setup(parser, args)
+    discover = _get_discover(parser, args, options)
+    interpreter = discover.interpreter
+    logging.debug("target interpreter %r", interpreter)
+    if interpreter is None:
+        raise RuntimeError("failed to find interpreter for {}".format(discover))
+    elements = [
+        _get_creator(interpreter, parser, options),
+        _get_seeder(parser, options),
+        _get_activation(interpreter, parser, options),
+    ]
+    [next(elem) for elem in elements]  # add choice of types
+    parser.parse_known_args(args, namespace=options)
+    [next(elem) for elem in elements]  # add type flags
```

Comment:
these list comprehensions seem very suspicious -- any way this can be done that's more explicit?